# Capstone Project for Coursera
This notebook will be mainly used for capstone project of Coursera online certificate for **IBM Data Science Professional Ceritificate**.

In [97]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
print("Hello Capstone Project Course!")

Hello Capstone Project Course!


### Import Libraries

First import the libraries which will be used in this capstone project.

In [98]:
import json

# !conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

import requests
from pandas.io.json import json_normalize

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

# !conda install -c conda-forge folium=0.5.0 --yes
import folium
print('Libraries imported.')

Libraries imported.


### Import Data

Next, using **BeautifulSoup** to transform the data in the table on the Wikipedia page into pandas dataframe. At the same time, cleanning dataframe as the requirement stated.

In [99]:
from bs4 import BeautifulSoup

with open('List of postal codes of Canada_ M - Wikipedia.html') as html_file:
    soup = BeautifulSoup(html_file, 'lxml')


In [174]:
t = soup.find('table', class_='wikitable sortable')
table_rows = table.find_all("tr")
table_columns = table.find_all("th")
l = []
for tr in table_rows:
    td = tr.find_all("td")
    row = [tr.text for tr in td]
    if row and row[1] != 'Not assigned':
        if row[2] == 'Not assigned\n':
            row[2] = row[1]
        l.append(row)
cols = [c.text.replace('\n', '') for c in table_columns]
df = pd.DataFrame(l, columns=cols).replace('\n', '', regex=True)
df = df.groupby(['Borough', 'Postcode'], sort=False).agg(lambda x: ', '.join(set(x.dropna()))).reset_index()
df = df[cols].replace('\n', '', regex=True)
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                                         Queen's Park  
5                                     Islington Avenue  
6                                       Rouge, Malvern  
7          

In [163]:
df.shape

(103, 3)

As the result of **shape** method, the size of this transformed table is **(103, 3)**.

Next step is add the **Latitude** and **Longitude** to the dataframe.

In [175]:
ll = pd.read_csv('Geospatial_Coordinates.csv')
ll.columns = ['Postcode', 'Latitude', 'Longitude']
cols = ['Postcode']
df_withll = df.join(ll.set_index(cols), on=cols)

### Use geopy library to get the latitude and longitude values of Toronto

In [184]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ca_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


### Create a map of Ontario with neighborhoods superimposed on top

In [185]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_withll['Latitude'], df_withll['Longitude'], df_withll['Borough'], df_withll['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Then filtering the **Borough** by contains **Toronto**

In [173]:
df_tor = df_withll[df_withll['Borough'].str.contains("Toronto")]
df_tor

Postcode           Borough  \
2        M5A  Downtown Toronto   
9        M5B  Downtown Toronto   
15       M5C  Downtown Toronto   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
54       M4M      East Toronto   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   

                                         Neighbourhood   Latitude  Longitude  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
9                             Ryerson, Garden District  43.657162 -79.378937  
15                                      St. James Town  43.651494 -79.375418  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
24                                  Central Bay Street  43.657952 -79.387383  
25                                            Christie  43.669542 -79.422564  
30                            King, Adelaide, Richmond  43.650571 -79.384568  
31                        Dovercourt Village, Dufferin  43.669005 -79.442259  
36   Toronto Islands, Harbourfront East, Union Station  43.640816 -79.381752  
37                            Little Portugal, Trinity  43.647927 -79.419750  
41                        The Danforth West, Riverdale  43.679557 -79.352188  
42            Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
43        Exhibition Place, Brockton, Parkdale Village  43.636847 -79.428191  
47                      India Bazaar, The Beaches West  43.668999 -79.315572  
48                      Commerce Court, Victoria Hotel  43.648198 -79.379817  
54                                     Studio District  43.659526 -79.340923  
61                                       Lawrence Park  43.728020 -79.388790  
62                                            Roselawn  43.711695 -79.416936  
67                                    Davisville North  43.712751 -79.390197  
68                 Forest Hill West, Forest Hill North  43.696948 -79.411307  
69                       The Junction South, High Park  43.661608 -79.464763  
73                                  North Toronto West  43.715383 -79.405678  
74                 North Midtown, The Annex, Yorkville  43.672710 -79.405678  
75                              Roncesvalles, Parkdale  43.648960 -79.456325  
79                                          Davisville  43.704324 -79.388790  
80                      Harbord, University of Toronto  43.662696 -79.400049  
81                                  Runnymede, Swansea  43.651571 -79.484450  
83                         Summerhill East, Moore Park  43.689574 -79.383160  
84           Grange Park, Kensington Market, Chinatown  43.653206 -79.400049  
86   Rathnelly, South Hill, Forest Hill SE, Summerh...  43.686412 -79.400049  
87   Railway Lands, King and Spadina, Harbourfront ...  43.628947 -79.394420  
91                                            Rosedale  43.679563 -79.

Creating map of Toronto using only Borough that contains Toronto.

In [191]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_tor['Latitude'], df_tor['Longitude'], df_tor['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

### Define Foursquare Credentials and Version

In [192]:
CLIENT_ID = 'M5NXRZI5KMSVEMTB1FHXAKUUXF5ODQFPRDMJQEZ053TRIV4W' # your Foursquare ID
CLIENT_SECRET = 'HBRGKYXZTPLCSDZPS4HQEEGZFOTJ2LKL4N5XMURF0OW3URRB' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: M5NXRZI5KMSVEMTB1FHXAKUUXF5ODQFPRDMJQEZ053TRIV4W
CLIENT_SECRET:HBRGKYXZTPLCSDZPS4HQEEGZFOTJ2LKL4N5XMURF0OW3URRB
